cmd/pj-rehearse-metrics/pj-rehearse-metrics.go

```diff
@@ -0,0 +1,185 @@
+		return
+	}
+}
+
+func (s *scraper) scrapePr(prefix string) {
+	prNumber := path.Base(prefix)
+	jobPrefix := path.Join(prefix, rehearseJobName)
+	jobPrefix = fmt.Sprintf("%s/", jobPrefix) // Needed because path.Join() removes trailing slash
+
+	runs := s.bucket.Objects(s.ctx, &storage.Query{Prefix: jobPrefix, Delimiter: "/"})
+	for {
+		attrs, err := runs.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			fmt.Printf("Next() failed err=%v\n", err)
+			break
+		}
+
+		// If we see non-empty `Name`, we got a full object name = a file
+		// At this level, we only care about dirs, not files.
+		if attrs.Name != "" {
+			continue
+		}
+
+		s.scrapeRun(attrs.Prefix, prNumber)
+	}
+}
+
+type options struct {
+	cacheDir string
+}
+
+func gatherOptions() options {
+	o := options{}
+	fs := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+
+	fs.StringVar(&o.cacheDir, "cache-dir", "", "Path to a directory where scraped metrics data will be saved. If not provided, cache will not be used.")
+
+	if err := fs.Parse(os.Args[1:]); err != nil {
+		logrus.WithError(err).Fatal("Failed to parse options")
+	}
+	return o
+}
+
+func main() {
+	o := gatherOptions()
+	var err error
+
+	ctx := context.Background()
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to create client")
+	}
+
+	if o.cacheDir == "" {
+		o.cacheDir, err = ioutil.TempDir("", "")
+		if err != nil {
+			logrus.WithError(err).Fatal("Failed to create a temporary directory")
+		}
+		defer func() {
+			if err := os.RemoveAll(o.cacheDir); err != nil {
+				logrus.Warnf("Failed to clean up a temporary directory %s (err=%v)", o.cacheDir, err)
+			}
+		}()
+	}
+
+	bucket := client.Bucket(originCiBucket)
+	pullRequests := bucket.Objects(ctx, &storage.Query{
+		Prefix:    "pr-logs/pull/openshift_release/",
+		Delimiter: "/",
+	})
+
+	scraper := scraper{ctx: ctx, bucket: bucket, cacheDir: o.cacheDir}
+	prPrefixes := make(chan string, 100)
+	progress := make(chan string, 100)
+
+	sem := semaphore.NewWeighted(workersCount)
+
+	for w := 0; w < workersCount; w++ {
+		if err := sem.Acquire(ctx, 1); err != nil {
+			logrus.WithError(err).Error("Failed to acquire semaphore to spawn a worker")
+			break
+		}
+		go func() {
+			defer sem.Release(1)
+			for prefix := range prPrefixes {
+				scraper.scrapePr(prefix)
+				progress <- prefix
+			}
+		}()
+	}
+
+	go func() {
+		defer close(progress)
+		for {
+			attrs, err := pullRequests.Next()
+			if err == iterator.Done {
+				break
+			}
+			if err != nil {
+				fmt.Printf("Next() failed: err=%v\n", err)
+				break
+			}
+
+			if attrs.Name != "" {
+				continue
+			}
+
+			prPrefixes <- attrs.Prefix
+		}
+		close(prPrefixes)
+		if err := sem.Acquire(ctx, workersCount); err != nil {
+			logrus.WithError(err).Error("Failed to acquire semaphore to wait on workers to finish")
+		}
+	}()
+
+	counter := 0
+	for done := range progress {
```

Comment:
This is a bit clumsy, you could use a locked value or atomic updates instead

Reply:
You mean for the counter? Isn't it an overkill (locked value certainly is...) for what is basically a activity indicator?